### ISSUE TITLE
RFC: Quantized Variables

### ISSUE THREAD

**Issue by lgeiger** [larq/larq#306]
This Request for Comments (RFC) outlines a proposed addition to the API of `larq`. It would be good to get feedback on this, so please ask questions if I need to elaborate more in some areas. WIP implementation can be viewed at #312.

## Objective

`larq` has a concept of quantized layers that define a [quantized operation using quantized inputs and variables](https://larq.dev/guides/key-concepts/#quantized-layers). Quantized layers are currently the only official way to interact with `larq`. E.g. accessing the precision of quantized variables or inputs are cumbersome to get and require checking `layer.quantized_latent_weights` and `layer.quantizers`. This often requires downstream code that needs access to variables to rely on private methods. We can fix this by either introducing convenience functions and a stable API of quantized layers to access quantized variables or extend the API with the concept of a quantized variable. This document will elaborate on the latter since it allows new use cases for `larq` and gives users more fine grained control when using custom layers or no layers at all. This concept can be extended to `tf.Tensor` for quantized activations which we can discuss in a future RFC.

## Design Proposal

There are a few ways we could implement this functionality, for the initial proposal I closely followed the implementation of [`AutoCastVariable`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/mixed_precision/experimental/autocast_variable.py#L28) which only wraps `tf.Variable`.
While the variable acts like a quantized variable the TensorFlow data type will still be floating-point in order to be compatible with all TensorFlow operations. This is typically referred to as fake quantization.

```python
class QuantizedVariable(tf.Variable):
    def __init__(self, variable, quantizer=None, precision=None):
        """Creates an QuantizedVariable instance.

        # Arguments
        variable: A floating-point resource variable to wrap.
        quantizer: An optional quantizer to transform the floating-point
            variable to a fake quantized variable.
        precision: An optional integer defining the precision of the quantized
            variable. If `None`, `quantizer.precision` is used.
        """
        pass

    @property
    def precision(self):
        """Returns the integer precision of the quantized variable."""
        pass

    def latent_value(self):
        """Returns the latent value of the variable."""
        pass

    # Overload all operators to use the quantized version and delegate all assign
    # operations to the latent weight.
```

This variable would act like a quantized variable in the forward pass by overloading the default operators like `__add__`, `__mul__`, etc. but would delegate updates via `self.assign*` to the underlying latent variable. This would make reasoning about quantized variables a lot easier and would allow features like the model converter or model summary to have easy access to all information.

### How can we handle variable saving?

For the purpose of resuming training the latent variable should be saved to the checkpoint. For checkpoints we can set this behaviour by delegating `_gather_saveables_for_checkpoint` to the wrapped variable. Will this also work for Keras when saving to `.h5` which might call `get_weights` during saving?

### How should we handle true (fake) quantized variables like in Bop?

The most straight forward approach would be to keep the current design and use `quantizer=None` and `precision=1`, though an alternative would be to introduce an extra class.

### How can we support quantizers with trainable variables?

We currently don't have  a quantizer with trainable variables in `larq` core, but probably add one in the near future (e.g. 8-bit). It would be good to see how this would work with this approach. See also #89 

### Don't quantize the value of the variable by default

This would have the advantage that we won't run into any problems with variable saving or other side effect we might neglect. This would make usability a bit worse since we would need a context manager to define a scope within which the variable would return the quantized value.

### Should we wrap an existing `tf.Variable` or create a new one?

I think the approach of wrapping an existing `tf.Variable` is easier to implement and maintain. To me, it also doesn't have major downsides from an user perspective.

### Extend quantized layers with methods to return quantized versions of the variables

We could extend the `larq` layers to allow better access to the underlying variables. However I think ultimately this would be a worse abstraction than introducing a new variable type.

## Testing process

Since this touches a few Keras and TensorFlow internals we need very good testing for it. Below I highlighted some code paths that might have non obvious interactions with the `QuantizedVariable` and need extra care:

- Multi GPU training and `tf.DistributedVariable`
- Keras Float 16 training using `AutoCastVariable`
- TPUStrategy (will only be supported by TensorFlow 2.1+, so no current concern)
- Model checkpointing and saving (see above)
- Gradient computation in eager, functional and legacy graph mode

**Comment by leonoverweel** [comment]
Thanks for this write-up @lgeiger; overall I think the design is a good idea. A few thoughts/responses:

#### (1) `QuantizedVariable` signature 

For the `QuantizedVariable` signature, I'd propose changing it to making the `precision` a required argument. I think it's good to be explicit about this in code because (a) the name of a quantizer may not make it clear what its output precision is, and (b) it'd allow us to have an `__init__()` check for whether `precision == quantizer.precision`, which I think would be useful.

#### (2) Keras saving

> Will this also work for Keras when saving to `.h5` which might call `get_weights` during saving?

Do we know if this is the case?

One way to handle this could be adding an optional `quantized` keyword argument that defaults to `False` to things like `get_weight`, so that when it's called by code that's not aware of quantization (e.g., code that doesn't explicitly pass `quantized=True`), it'll just return the latent variable. A drawback is that this is a bit ugly and easy to forget to always do when writing quantized-aware code.

Another way could maybe be some kind of execution context? Where quantized variables only return their quantized values if they run in a `with quantized_variables:` context. I'm not too familiar with this so I'm not sure if it's a good idea. (Oh I see now that you suggested this a bit later in the RFC. I think it could be a good approach then.)

#### (3) True quantized variables

> How should we handle true (fake) quantized variables like in Bop?

I like the current design of just not setting a quantizer. I think the extra API surface/complexity of having e.g. a `QuantizedVariable` and a `TrulyQuantizedVariable` isn't worth it.

#### (4) Wrap `tf.Variable` or create a new one?

I agree that wrapping `tf.Variable` is good.

**Comment by lgeiger** [comment]
@leonoverweel Thanks for the comments.

> I think it's good to be explicit about this in code because

In general, I agree. But how would you call this from the [`larq.layers`](https://larq.dev/api/layers/#quantdense) API?

> Do we know if this is the case?

No I haven't checked, but I suspect it.

**Comment by leonoverweel** [comment]
How are variables being called from the `larq.layers` API now? Is that handled in `layers_base`?

**Comment by lgeiger** [comment]
> How are variables being called from the larq.layers API now? Is that handled in layers_base?

Yes, if we would require a precision argument, we probably also need to add that to larq.layers for complentess, which I think would make usability worse.

**Comment by leonoverweel** [comment]
These are the two options we have for integrating `QuantizedVariable` into the larger Larq API.

### Layer-level

Pass an optional precision to a `lq.layers.Layer`, which then uses it to set the `self.kernel` and `self.input` (?) for the layer to be `QuantizedVariable(precision=1)`.

```python
layer = lq.layers.QuantDense(
    32,
    input_precision=1,
    kernel_precision=1
)   # leaving quantizers as None
```

*Advantages:*
1. Explicit, intuitive from a user perspective.

*Disadvantages:*
1. Adds another set of key-word arguments to all layers: `input_precision` and `kernel_precision`.
1. What happens if you pass both a precision argument and a quantizer? Presumably, the quantizer would be set on the `QuantizedVariable`, but then it'll no longer get caught by e.g. Bop, which is unintuitive.

###  No-op Quantizer

Pass a `NoOpQuantizer`, which simply passes the value through, to a `lq.layers.Layer`, which then detects the precision of the `NoOpQuantizer` and uses it to set the precision of its variables.

```python
layer = lq.layers.QuantDense(
    32,
    input_quantizer=NoOpQuantizer(precision=1),
    kernel_quantizer=NoOpQuantizer(precision=1),
)
```

An open question is then to think of a better name than `NoOpOptimizer`.

*Advantages:*
1. Doesn't increase our layers API surface.
1. Less confusion because you can't pass another quantizer.

*Disadvantages:*
1. Passing a fake quantizer to set a variable's precision is less intuitive.

---

In both of these scenarios, Bop's check would become something like:

```python
@staticmethod
def is_binary_variable(var):
    return var.precision == 1 and var.quantizer is None
```

**Comment by jneeven** [comment]
I think that `NoOpQuantizer` would be the cleanest approach

**Comment by lgeiger** [comment]
@leonoverweel Thanks for outlining the different options.
I cannot come up with a good reason why anyone would want to use this feature to quantized input activations. The only thing that might be possible is if people want to quantize activations via a regularization loss. If we really don't have a good use case for it we could remove the need to introduce a `input_precision` argument which means that solution one only very slightly increases the API surface of the layers.

In any case, I am still slightly in favour of the "int quantizer" solution due to my obsession with short syntax. Though I see why this might be confusing so the `NoOpQuantizer` is probably a more sensible choice.

**Comment by leonoverweel** [comment]
I think if we document it well, I'd be good with going for the int quantizer approach.

**Comment by lgeiger** [comment]
Thinking about this again, I think the `NoOpQuantizer` quantizer would be the cleanest solution and we can even add metrics to it following #343

**Comment by koenhelwegen** [comment]
Hi, my apologies for not thinking this through earlier but after seeing the PR I am concerned this approach will make Larq more opaque for people who are not already very familiar with BNNs, STEs and Bop.

In particular, I'm worried that the possibility of setting `kernel_quantizer=1`, which effectively results in `kernel_quantizer=None`, will result in people thinking they have binary models while really training full precision models. This is worse if they superficially know about the `precision` attribute and check that to see if they are working with 1-bit precision.

**Default quantizer**
However, I do like this syntax. What I would propose is the following:
- default to a quantizer `SteSign(clip_value=1.1)` and a constraint `WeightClip(clip_value=1.1)` if the user sets `kernel_quantizer=1`,
- use similar defaults (or throw an error) for other integers,
- and make the default check for Bop `var.precision == 1`.

This will result in intuitive behaviour for both Adam/SGD/etc and Bop, and keeps syntax short. The disadvantage is it will result in a performance degradation, which we should check (although some initial comparisons on my local machine suggest this is very small/negligible). We could add a `NoOpQuantizer` as well for sophisticated users concerned about this (although I suppose that would not be very pythonic).

Note that sophisticated users still can pass an arbitrary quantizer and implement any behaviour they like.

I'm also fine with dropping the integer notation and using the `NoOpQuantizer` instead - that way the lack of quantization is explicit.

**Comment by lgeiger** [comment]
I see your concerns  and agree that `kernel_quantizer=1` can be confusing if one doesn't know about the implementation. I also think adding an additional keyword argument like `kernel_precision` will be problematic since one would expect the layer to handle quantization then, where as it is done via the optimizer.

I am a bit reluctant to adding a default quantizer, since for me there is no obvious default choice especially for other bit-widths. Default choices like this are hard to change in the future. E.g. there is a lot of people asking to change the default Keras kernel initializer for convolutions to something like `he_normal` which is tricky to do since it would break a lot of users relying on the default being `glorot_uniform`.

> I'm also fine with dropping the integer notation and using the NoOpQuantizer instead - that way the lack of quantization is explicit.

I think I would actually prefer this option. Though I am not a big fan of the `NoOpQuantizer` name. What do you think about something like `MarkExternalQuantizer`, `TagExternalQuantizer`, `PseudoQuantizer` which should hint that the quantization is expected to be handled outside of the layer (i.e. variables are replaced with their quantized values by the user or an external optimizer like `Bop`)? We can then add further options to the `NoopQuantizer` to do some assertions (e.g. check that values are actually -1 or 1).

@koenhelwegen @leonoverweel @jneeven What do you think about this from a user perspective?

**Comment by leonoverweel** [comment]
I agree with your point about adding a default quantizer. I also don't see an obvious candidate for a default now, and it could indeed be tricky if we realize the default we end up picking isn't the most sensible one in the future.

The `NoOpQuantizer` is probably the best way to go, yeah. Adding asserts would be cool. For naming, I'd suggest `PassQuantizer` or `QuantizerPass`: it is reminiscent of Python's `pass` statement (implying that it is a placeholder with no inherent functionality) and hints that the responsibility of quantizing is "passed on" to another part of the codebase (such as the optimizer).

Dropping the integer notation will also simplify the API, so I'm in favor.

**Comment by lgeiger** [comment]
> For naming, I'd suggest `PassQuantizer` or `QuantizerPass`

I like the reasoning behind this name, though it reminds me of a compiler pass instead of Python's `pass` statement. E.g. during the conversion to a TFLite model, TensorFlow uses a [MLIR QuantizationPass](https://github.com/tensorflow/tensorflow/blob/2938772a08ed02ced4663ca38168ab3f82e8f81b/tensorflow/compiler/mlir/lite/tf_tfl_passes.h#L30-L33) to do post-training quantizations.

> Dropping the integer notation will also simplify the API, so I'm in favor.

:+1:

**Comment by leonoverweel** [comment]
>  E.g. during the conversion to a TFLite model, TensorFlow uses a MLIR QuantizationPass to do post-training quantizations.

Maybe using the `PassQuantizer` order (instead of `QuantizerPass`) will be enough differentiation?

**Comment by lgeiger** [comment]
> Maybe using the `PassQuantizer` order (instead of `QuantizerPass`) will be enough differentiation?

What about `PassThroughQuantizer`? Though that feels quite verbose

**Comment by jneeven** [comment]
So why exactly is `NoOpQuantizer` not a suitable name? Seems to describe its workings pretty accurately to me.

**Comment by lgeiger** [comment]
> So why exactly is `NoOpQuantizer` not a suitable name

Good point. I am beginning to like the `NoOpQuantizer` (or `NoopQuantizer`) name again. I like `FakeQuantizer` as well, though it might get confused with the [TensorFlow fake quantizers](https://www.tensorflow.org/api_docs/python/tf/quantization/fake_quant_with_min_max_args).

**Comment by jneeven** [comment]
> > So why exactly is `NoOpQuantizer` not a suitable name
> 
> Good point. I am beginning to like the `NoOpQuantizer` (or `NoopQuantizer`) name again. I like `FakeQuantizer` as well, though it might get confused with the [TensorFlow fake quantizers](https://www.tensorflow.org/api_docs/python/tf/quantization/fake_quant_with_min_max_args).

I'd say the fault there is with TF for picking a shitty name, but indeed let's not use `FakeQuantizer` to prevent confusion then.

**Comment by lgeiger** [comment]
It looks like we all seem to prefer going for a `NoOpQuantizer` approach and we are only debating the name at this point.
To summarise, the usage for an optimizer like Bop would be:
```python
layer = lq.layers.QuantDense(
    32,
    input_quantizer=lq.quantizers.SteSign(),
    kernel_quantizer=lq.quantizers.NoOpQuantizer(precision=1),
)
```
Here, `NoOpQuantizer` doesn't do any quantization and only sets `precision=1` on the variable and will include the relevant training metrics (see #402).

The `is_binary` check in `Bop` would change to:
```python
@staticmethod
def is_binary_variable(var):
    return var.precision == 1 and var.quantizer is None
```

@larq/core I am still very open to more suggestions on how to do this in a better way, but it would be good to come to an decision soon.

**Comment by leonoverweel** [comment]
@lgeiger yes, I agree with this. But wouldn't the `is_binary` check in `Bop` be more like this?

```python
@staticmethod
def is_binary_variable(var):
    return var.precision == 1 and isinstance(var.quantizer, NoOpQuantizer)
```

**Comment by MariaHeuss** [comment]
From the point of few of a heavy Bop user: I like the NoOpQuantizer approach, I am in favor of NoOpQuantizer or NoQuantizer as name, FakeQuantizer is too confusing indeed. It looks pretty intuitive to me also and I can not think of something I want to do with Bop that would not work with this approach.

**Comment by lgeiger** [comment]
Closed via #442
